azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -273,6 +273,109 @@ def _handle_pipeline_event(self, event):
+        # a few seconds given processing time. We could make this more accurate if SasToken
+        # objects had a live TTL value rather than a static one (i.e. "there are n seconds
+        # remaining in the lifespan of this token", rather than "this token was intended to live
+        # for n seconds")
+        seconds_until_renewal = (
+            self.pipeline_root.pipeline_configuration.sastoken.ttl
+            - self.DEFAULT_TOKEN_RENEWAL_MARGIN
+        )
+        if seconds_until_renewal < 0:
+            # This shouldn't happen in correct flow, but it's possible I suppose, if something goes
+            # horribly awry elsewhere in the stack, or if we start allowing for custom
+            # SasToken TTLs or custom Renewal Margins in the future
+            logger.error("ERROR: SasToken TTL less than Renewal Margin")
+            handle_exceptions.handle_background_exception(
+                pipeline_exceptions.PipelineError("SasToken TTL less than Renewal Margin!")
+            )
+        else:
+            logger.debug(
+                "Scheduling SAS Token renewal for {} seconds in the future".format(
+                    seconds_until_renewal
+                )
+            )
+            self_weakref = weakref.ref(self)
+
+            @pipeline_thread.runs_on_pipeline_thread
+            def on_reauthorize_complete(op, error):
+                this = self_weakref()
+                if error:
+                    logger.error(
+                        "{}({}): reauthorize connection operation failed.  Error={}".format(
+                            this.name, op.name, error
+                        )
+                    )
+                    handle_exceptions.handle_background_exception(error)
+                else:
+                    logger.debug(
+                        "{}({}): reauthorize connection operation is complete".format(
+                            this.name, op.name
+                        )
+                    )
+
+            @pipeline_thread.invoke_on_pipeline_thread_nowait
+            def renew_token():
+                this = self_weakref()
+                logger.debug("Renewing SAS Token")
+                # Renew the token
+                sastoken = this.pipeline_root.pipeline_configuration.sastoken
+                sastoken.refresh()
+                # If the pipeline is already connected, send order to reauthorize the connection
+                # now that token has been renewed
+                if this.pipeline_root.connected:
+                    this.send_op_down(
+                        pipeline_ops_base.ReauthorizeConnectionOperation(
+                            callback=on_reauthorize_complete
+                        )
+                    )
+                # Once again, start a renewal timer
+                this._start_renewal_timer()
+
+            self._token_renewal_timer = threading.Timer(seconds_until_renewal, renew_token)
+            self._token_renewal_timer.daemon = True
```

Comment:
>self._token_renewal_timer.daemon = True [](start = 12, length = 39)

Mental exercise: Consider different token lifetime rules.  why do we create a token on pipeline init and not on connect?  Why do we keep the token renewal timer going after we disconnect?  We don't need the token before we connect and we don't need it after we disconnect. 

This is why I'm asking the question: daemon=True protects us for shutdown scenarios.  In other words, if a daemon timer (thread) is still running when we shut down, it won't stop the app from exiting.  That's good.  

But, daemon=True doesn't stop the timer from living longer than the pipeline.  Consider an app that makes multiple clients, with each client sending a message then dying.  For each client instance, there would be a thread that lives for up to a full hour past the client.  Then, when each timer fires, it silently raises an exception when it tries to dereference `this` and dies.  No harm, but still sloppy and sort-of a memory leak.